Trips are sometimes incorrectly "joined" even if the gap between sections is close to an hour long

We have just started our trip, yay! Now we still don't have enough points to make a decision.

```
2016-08-15 12:10:17,370:DEBUG:------------------------------2016-08-10T09:54:31-07:00------------------------------
2016-08-15 12:10:17,372:DEBUG:last5MinsDistances = [ 52.82764235] with length 1
2016-08-15 12:10:17,374:DEBUG:last10PointsDistances = [ 52.82764235   0.        ] with length 2, shape (2,)
2016-08-15 12:10:17,374:DEBUG:len(last10PointsDistances) = 2, len(last5MinsDistances) = 1
2016-08-15 12:10:17,375:DEBUG:last5MinsTimes.max() = 31.5750000477, time_threshold = 300
2016-08-15 12:10:17,375:DEBUG:Too few points to make a decision, continuing
...
2016-08-15 12:10:17,435:DEBUG:------------------------------2016-08-10T09:58:13-07:00------------------------------
2016-08-15 12:10:17,440:DEBUG:last5MinsDistances = [ 329.2908086   280.89265787  245.2010695   184.71648394  189.02429048
  142.74543776   87.77772771   43.01575703   56.99527453] with length 9
2016-08-15 12:10:17,444:DEBUG:last10PointsDistances = [ 329.2908086   280.89265787  245.2010695   184.71648394  189.02429048
  142.74543776   87.77772771   43.01575703   56.99527453    0.        ] with length 10, shape (10,)
2016-08-15 12:10:17,446:DEBUG:len(last10PointsDistances) = 10, len(last5MinsDistances) = 9
2016-08-15 12:10:17,446:DEBUG:last5MinsTimes.max() = 253.575000048, time_threshold = 300
2016-08-15 12:10:17,447:DEBUG:Too few points to make a decision, continuing
```

Now we finally have enough points

```
2016-08-15 12:10:17,447:DEBUG:------------------------------2016-08-10T09:58:52-07:00------------------------------
2016-08-15 12:10:17,454:DEBUG:last5MinsDistances = [ 368.82243137  323.60485982  289.21766655  230.01780163  235.70421315
  192.31601923  144.97574662   96.48072959   89.58728643   59.109117  ] with length 10
2016-08-15 12:10:17,457:DEBUG:last10PointsDistances = [ 323.60485982  289.21766655  230.01780163  235.70421315  192.31601923
  144.97574662   96.48072959   89.58728643   59.109117      0.        ] with length 10, shape (10,)
2016-08-15 12:10:17,460:DEBUG:len(last10PointsDistances) = 10, len(last5MinsDistances) = 10
2016-08-15 12:10:17,461:DEBUG:last5MinsTimes.max() = 292.575000048, time_threshold = 300
2016-08-15 12:10:17,461:DEBUG:last5MinsDistances.max() = 368.822431372, last10PointsDistance.max() = 323.604859817
...
2016-08-15 12:10:17,609:DEBUG:------------------------------2016-08-10T10:03:53-07:00------------------------------
2016-08-15 12:10:17,615:DEBUG:last5MinsDistances = [ 131.22779511  122.19976408   77.28623256   51.39123828   36.04130693
   39.82444342   25.54539197   14.229786     34.26178107   25.54539197
   36.07236258] with length 11
2016-08-15 12:10:17,619:DEBUG:last10PointsDistances = [ 77.28623256  51.39123828  36.04130693  39.82444342  25.54539197
  14.229786    34.26178107  25.54539197  36.07236258   0.        ] with length 10, shape (10,)
2016-08-15 12:10:17,622:DEBUG:len(last10PointsDistances) = 10, len(last5MinsDistances) = 11
2016-08-15 12:10:17,622:DEBUG:last5MinsTimes.max() = 271.786000013, time_threshold = 300
2016-08-15 12:10:17,622:DEBUG:last5MinsDistances.max() = 131.227795112, last10PointsDistance.max() = 77.2862325602
```

Here's where we should segment. Here, we have enough points, and the
last10PointsDistance.max() is only 77, but the last 5 mins distances > 100

And then we only get a point on the return trip.

```
2016-08-15 12:10:17,623:DEBUG:------------------------------2016-08-10T11:07:05.684000-07:00------------------------------
2016-08-15 12:10:17,625:DEBUG:last5MinsDistances = [] with length 0
2016-08-15 12:10:17,629:DEBUG:last10PointsDistances = [ 109.66159855  129.1706589   118.45003771  131.19699388  146.88650462
  122.53046402  131.19699388  120.7937173   156.72206388    0.        ] with length 10, shape (10,)
2016-08-15 12:10:17,629:DEBUG:len(last10PointsDistances) = 10, len(last5MinsDistances) = 0
```

Now testing with the trip end detection missing.
As expected, the test fails because we skip the trip whose trip end was detected.

on android

```
2016-08-21 07:54:07,180:DEBUG:140735159468032:Comparing 2016-08-09T17:54:54.529000-07:00 -> 2016-08-09T18:00:46-07:00 with 2016-08-09T17:54:54.529000-07:00 -> 2016-08-09T18:00:46-07:00
2016-08-21 07:54:07,181:DEBUG:140735159468032:Comparing 2016-08-09T18:04:16.449000-07:00 -> 2016-08-09T18:20:41-07:00 with 2016-08-09T18:04:16.449000-07:00 -> 2016-08-09T18:20:41-07:00
2016-08-21 07:54:07,181:DEBUG:140735159468032:Comparing 2016-08-10T11:07:05.684000-07:00 -> 2016-08-10T11:14:45.512000-07:00 with 2016-08-10T09:53:59.425000-07:00 -> 2016-08-10T10:02:11-07:00
2016-08-21 07:54:07,181:DEBUG:140735159468032:Comparing 2016-08-10T17:43:37.535000-07:00 -> 2016-08-10T18:00:18.899000-07:00 with 2016-08-10T11:07:05.684000-07:00 -> 2016-08-10T11:14:45.512000-07:00
2016-08-21 07:54:07,181:DEBUG:140735159468032:Comparing 2016-08-10T18:11:04.989000-07:00 -> 2016-08-10T18:13:49.871000-07:00 with 2016-08-10T17:43:37.535000-07:00 -> 2016-08-10T18:00:18.899000-07:00
2016-08-21 07:54:07,182:DEBUG:140735159468032:Comparing 2016-08-10T19:28:06.757000-07:00 -> 2016-08-10T19:48:43.490000-07:00 with 2016-08-10T18:11:04.989000-07:00 -> 2016-08-10T18:13:49.871000-07:00
F2016-08-21 07:54:07,182:DEBUG:140735159468032:Clearing related databases

======================================================================
FAIL: testAug10MultiSyncEndNotDetected (__main__.TestPipelineRealData)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/analysisTests/intakeTests/TestPipelineRealData.py", line 349, in testAug10MultiSyncEndNotDetected
    ad.AttrDict(ground_truth).data, time_fuzz=60, distance_fuzz=100)
  File "emission/tests/analysisTests/intakeTests/TestPipelineRealData.py", line 92, in compare_approx_result
    self.assertEqual(len(result), len(expect))
AssertionError: 6 != 7

----------------------------------------------------------------------
```
and iOS

```
2016-08-21 08:08:16,808:DEBUG:140735159468032:Comparing 2016-02-22T13:06:06.142961-08:00 -> 2016-02-22T13:13:19.001847-08:00 with 2016-02-22T13:06:06.142961-08:00 -> 2016-02-22T13:13:19.001847-08:00
2016-08-21 08:08:16,808:DEBUG:140735159468032:Comparing 2016-02-22T15:33:18.550963-08:00 -> 2016-02-22T15:37:12.000136-08:00 with 2016-02-22T15:33:18.550963-08:00 -> 2016-02-22T15:37:12.000136-08:00
2016-08-21 08:08:16,809:DEBUG:140735159468032:Comparing 2016-02-22T15:49:43.999972-08:00 -> 2016-02-22T16:04:00.997765-08:00 with 2016-02-22T15:49:43.999972-08:00 -> 2016-02-22T16:04:00.997765-08:00
2016-08-21 08:08:16,809:DEBUG:140735159468032:Comparing 2016-02-22T19:41:42.085297-08:00 -> 2016-02-22T19:48:35.000033-08:00 with 2016-02-22T16:49:30.999990-08:00 -> 2016-02-22T16:56:58.000047-08:00
2016-08-21 08:08:16,809:DEBUG:140735159468032:Comparing 2016-02-22T19:58:50.000111-08:00 -> 2016-02-22T20:12:15.000095-08:00 with 2016-02-22T19:41:42.085297-08:00 -> 2016-02-22T19:48:35.000033-08:00
F2016-08-21 08:08:16,810:DEBUG:140735159468032:Clearing related databases

======================================================================
FAIL: testFeb22MultiSyncEndNotDetected (__main__.TestPipelineRealData)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "emission/tests/analysisTests/intakeTests/TestPipelineRealData.py", line 387, in testFeb22MultiSyncEndNotDetected
    ad.AttrDict(ground_truth).data, time_fuzz=60, distance_fuzz=100)
  File "emission/tests/analysisTests/intakeTests/TestPipelineRealData.py", line 92, in compare_approx_result
    self.assertEqual(len(result), len(expect))
AssertionError: 5 != 6

----------------------------------------------------------------------
```

--- a/emission/analysis/intake/segmentation/trip_segmentation_methods/dwell_segmentation_dist_filter.py
+++ b/emission/analysis/intake/segmentation/trip_segmentation_methods/dwell_segmentation_dist_filter.py
@@ -36,14 +36,9 @@ class DwellSegmentationDistFilter(eaist.TripSegmentationMethod):
         """
         filtered_points_df = timeseries.get_data_df("background/filtered_location", time_query)
         transition_df = timeseries.get_data_df("statemachine/transition", time_query)
+        logging.debug("transition_df = %s" % transition_df[["fmt_time", "transition"]])
 
-        if len(filtered_points_df) == 0:
-            self.last_ts_processed = None
-        else:
-            # TODO: Decide whether we should return the write_ts in the entry,
-            # or whether we should search by timestamp instead.
-            # Depends on final direction for the timequery
-            self.last_ts_processed = filtered_points_df.iloc[-1].metadata_write_ts
+        self.last_ts_processed = None
 
         logging.info("Last ts processed = %s" % self.last_ts_processed)
 
@@ -61,6 +56,8 @@ class DwellSegmentationDistFilter(eaist.TripSegmentationMethod):
 
             if just_ended:
                 if self.continue_just_ended(idx, currPoint, filtered_points_df):
+                    # We have "processed" the currPoint by deciding to glom it
+                    self.last_ts_processed = currPoint.ts
                     continue
                 # else: 
                 # Here's where we deal with the start trip. At this point, the
@@ -94,6 +91,8 @@ class DwellSegmentationDistFilter(eaist.TripSegmentationMethod):
                         (last_trip_end_point, idx-1))
                     segmentation_points.append((curr_trip_start_point, last_trip_end_point))
                     logging.info("Found trip end at %s" % last_trip_end_point.fmt_time)
+                    # We have processed everything up to the trip end by marking it as a completed trip
+                    self.last_ts_processed = currPoint.ts
                     just_ended = True
                     # Now, we have finished processing the previous point as a trip
                     # end or not. But we still need to process this point by seeing
@@ -133,9 +132,11 @@ class DwellSegmentationDistFilter(eaist.TripSegmentationMethod):
         # then we end the trip at the last point that we have.
         if not just_ended and len(transition_df) > 0:
             stopped_moving_after_last = transition_df[(transition_df.ts > currPoint.ts) & (transition_df.transition == 2)]
+            logging.debug("stopped_moving_after_last = %s" % stopped_moving_after_last[["fmt_time", "transition"]])
             if len(stopped_moving_after_last) > 0:
                 logging.debug("Found %d transitions after last point, ending trip..." % len(stopped_moving_after_last))
                 segmentation_points.append((curr_trip_start_point, currPoint))
+                self.last_ts_processed = currPoint.ts
             else:
                 logging.debug("Found %d transitions after last point, not ending trip..." % len(stopped_moving_after_last))
         return segmentation_points

--- a/emission/analysis/intake/segmentation/trip_segmentation_methods/dwell_segmentation_time_filter.py
+++ b/emission/analysis/intake/segmentation/trip_segmentation_methods/dwell_segmentation_time_filter.py
@@ -52,15 +52,9 @@ class DwellSegmentationTimeFilter(eaist.TripSegmentationMethod):
         """
         filtered_points_df = timeseries.get_data_df("background/filtered_location", time_query)
         transition_df = timeseries.get_data_df("statemachine/transition", time_query)
-        logging.debug("transition_df = %s" % transition_df)
+        logging.debug("transition_df = %s" % transition_df[["fmt_time", "transition"]])
 
-        if len(filtered_points_df) == 0:
-            self.last_ts_processed = None
-        else:
-            # TODO: Decide whether we should return the write_ts in the entry,
-            # or whether we should search by timestamp instead.
-            # Depends on final direction for the timequery
-            self.last_ts_processed = filtered_points_df.iloc[-1].metadata_write_ts
+        self.last_ts_processed = None
 
         logging.info("Last ts processed = %s" % self.last_ts_processed)
 
@@ -79,6 +73,8 @@ class DwellSegmentationTimeFilter(eaist.TripSegmentationMethod):
 
             if just_ended:
                 if self.continue_just_ended(idx, currPoint, filtered_points_df):
+                    # We have "processed" the currPoint by deciding to glom it
+                    self.last_ts_processed = currPoint.ts
                     continue
                 # else:
                 sel_point = currPoint
@@ -119,6 +115,8 @@ class DwellSegmentationTimeFilter(eaist.TripSegmentationMethod):
                                                                                        last10Points_df, last5MinsPoints_df)
                 segmentation_points.append((curr_trip_start_point, last_trip_end_point))
                 logging.info("Found trip end at %s" % last_trip_end_point.fmt_time)
+                # We have processed everything up to the trip end by marking it as a completed trip
+                self.last_ts_processed = currPoint.ts
                 if ended_before_this:
                     # in this case, we end a trip at the previous point, and the next trip starts at this
                     # point, not the next one
@@ -140,6 +138,9 @@ class DwellSegmentationTimeFilter(eaist.TripSegmentationMethod):
                 (unused, last_trip_end_point) = self.get_last_trip_end_point(filtered_points_df,
                                                                              last10Points_df, None)
                 segmentation_points.append((curr_trip_start_point, last_trip_end_point))
+                logging.debug("Found trip end at %s" % last_trip_end_point.fmt_time)
+                # We have processed everything up to the trip end by marking it as a completed trip
+                self.last_ts_processed = currPoint.ts
 
         return segmentation_points
 

--- a/emission/tests/analysisTests/intakeTests/TestPipelineRealData.py
+++ b/emission/tests/analysisTests/intakeTests/TestPipelineRealData.py
@@ -308,6 +308,83 @@ class TestPipelineRealData(unittest.TestCase):
         self.compare_approx_result(ad.AttrDict({'result': api_result}).result,
                                    ad.AttrDict(ground_truth).data, time_fuzz=60, distance_fuzz=100)
 
+    def testAug10MultiSyncEndNotDetected(self):
+        # Re-run, but with multiple calls to sync data
+        # This tests the effect of online versus offline analysis and segmentation with potentially partial data
+
+        dataFile = "emission/tests/data/real_examples/shankari_2016-08-10"
+        start_ld = ecwl.LocalDate({'year': 2016, 'month': 8, 'day': 9})
+        end_ld = ecwl.LocalDate({'year': 2016, 'month': 8, 'day': 10})
+        cacheKey = "diary/trips-2016-08-10"
+        ground_truth = json.load(open("emission/tests/data/real_examples/shankari_2016-08-910.ground_truth"),
+                                 object_hook=bju.object_hook)
+
+        logging.info("Before loading, timeseries db size = %s" % edb.get_timeseries_db().count())
+        all_entries = json.load(open(dataFile), object_hook = bju.object_hook)
+        ts_1030 = arrow.get("2016-08-10T10:30:00-07:00").timestamp
+        logging.debug("ts_1030 = %s, converted back = %s" % (ts_1030, arrow.get(ts_1030).to("America/Los_Angeles")))
+        before_1030_entries = [e for e in all_entries if ad.AttrDict(e).metadata.write_ts <= ts_1030]
+        after_1030_entries = [e for e in all_entries if ad.AttrDict(e).metadata.write_ts > ts_1030]
+
+        # First load all data from the 9th. Otherwise, the missed trip is the first trip,
+        # and we don't set the last_ts_processed
+        # See the code around "logging.debug("len(segmentation_points) == 0, early return")"
+        etc.setupRealExample(self, "emission/tests/data/real_examples/shankari_2016-08-09")
+
+        # Sync at 10:30 to capture all the points on the trip *to* the optometrist
+        # Skip the last few points to ensure that the trip end is skipped
+        self.entries = before_1030_entries[0:-2]
+        etc.setupRealExampleWithEntries(self)
+        etc.runIntakePipeline(self.testUUID)
+        api_result = gfc.get_geojson_for_dt(self.testUUID, start_ld, end_ld)
+
+        # Then sync after 10:30
+        self.entries = after_1030_entries
+        etc.setupRealExampleWithEntries(self)
+        etc.runIntakePipeline(self.testUUID)
+        api_result = gfc.get_geojson_for_dt(self.testUUID, start_ld, end_ld)
+
+        # Although we process the day's data in two batches, we should get the same result
+        self.compare_approx_result(ad.AttrDict({'result': api_result}).result,
+                                   ad.AttrDict(ground_truth).data, time_fuzz=60, distance_fuzz=100)
+
+    def testFeb22MultiSyncEndNotDetected(self):
+        # Re-run, but with multiple calls to sync data
+        # This tests the effect of online versus offline analysis and segmentation with potentially partial data
+
+        dataFile = "emission/tests/data/real_examples/iphone_2016-02-22"
+        start_ld = ecwl.LocalDate({'year': 2016, 'month': 2, 'day': 22})
+        end_ld = ecwl.LocalDate({'year': 2016, 'month': 2, 'day': 22})
+        cacheKey = "diary/trips-2016-02-22"
+        ground_truth = json.load(open(dataFile+".ground_truth"), object_hook=bju.object_hook)
+
+        logging.info("Before loading, timeseries db size = %s" % edb.get_timeseries_db().count())
+        all_entries = json.load(open(dataFile), object_hook = bju.object_hook)
+        # 18:01 because the transition was at 2016-02-22T18:00:09.623404-08:00, so right after
+        # 18:00
+        ts_1800 = arrow.get("2016-02-22T18:00:30-08:00").timestamp
+        logging.debug("ts_1800 = %s, converted back = %s" % (ts_1800, arrow.get(ts_1800).to("America/Los_Angeles")))
+        before_1800_entries = [e for e in all_entries if ad.AttrDict(e).metadata.write_ts <= ts_1800]
+        after_1800_entries = [e for e in all_entries if ad.AttrDict(e).metadata.write_ts > ts_1800]
+
+        # Sync at 18:00 to capture all the points on the trip *to* the optometrist
+        # Skip the last few points to ensure that the trip end is skipped
+        import uuid
+        self.testUUID = uuid.uuid4()
+        self.entries = before_1800_entries[0:-2]
+        etc.setupRealExampleWithEntries(self)
+        etc.runIntakePipeline(self.testUUID)
+        api_result = gfc.get_geojson_for_dt(self.testUUID, start_ld, end_ld)
+
+        # Then sync after 18:00
+        self.entries = after_1800_entries
+        etc.setupRealExampleWithEntries(self)
+        etc.runIntakePipeline(self.testUUID)
+        api_result = gfc.get_geojson_for_dt(self.testUUID, start_ld, end_ld)
+
+        # Although we process the day's data in two batches, we should get the same result
+        self.compare_approx_result(ad.AttrDict({'result': api_result}).result,
+                                   ad.AttrDict(ground_truth).data, time_fuzz=60, distance_fuzz=100)
 
 if __name__ == '__main__':
     etc.configLogging()

